pkg/reconciler/events/scheduler/scheduler.go

```diff
@@ -0,0 +1,353 @@
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := r.updateStatus(ctx, scheduler); uErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update Scheduler status", zap.Error(uErr))
+		r.Recorder.Eventf(scheduler, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for Scheduler %q: %v", scheduler.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(scheduler, corev1.EventTypeNormal, "Updated", "Updated Scheduler %q", scheduler.Name)
+	}
+	if reconcileErr != nil {
+		r.Recorder.Event(scheduler, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, scheduler *v1alpha1.Scheduler) error {
+	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("scheduler", scheduler)))
+
+	// If jobName / topic has been already configured, stash them here
+	// since right below we remove them.
+	jobName := scheduler.Status.JobName
+	topic := scheduler.Status.TopicID
+
+	scheduler.Status.InitializeConditions()
+
+	// And restore them.
+	scheduler.Status.JobName = jobName
+	if topic == "" {
+		topic = resources.GenerateTopicName(scheduler)
+	}
+
+	// See if the source has been deleted.
+	if scheduler.DeletionTimestamp != nil {
+		logging.FromContext(ctx).Desugar().Debug("Deleting Scheduler job")
+		if err := r.deleteJob(ctx, scheduler); err != nil {
+			scheduler.Status.MarkJobNotReady("JobDeleteFailed", "Failed to delete Scheduler job: %s", err.Error())
+			return err
+		}
+		scheduler.Status.MarkJobNotReady("JobDeleted", "Successfully deleted Scheduler job: %s", scheduler.Status.JobName)
+		scheduler.Status.JobName = ""
+
+		if err := r.PubSubBase.DeletePubSub(ctx, scheduler); err != nil {
+			return err
+		}
+		removeFinalizer(scheduler)
+		return nil
+	}
+
+	// Ensure that there's finalizer there, since we're about to attempt to
+	// change external state with the topic, so we need to clean it up.
+	addFinalizer(scheduler)
+
+	_, _, err := r.PubSubBase.ReconcilePubSub(ctx, scheduler, topic, resourceGroup)
+	if err != nil {
+		return err
+	}
+
+	jobName, err = r.reconcileJob(ctx, scheduler, topic, resources.GenerateJobName(scheduler))
+	if err != nil {
+		scheduler.Status.MarkJobNotReady("JobCreateFailed", "Failed to create Scheduler job: %s", err.Error())
+		return err
+	}
+	scheduler.Status.MarkJobReady(jobName)
+	return nil
+}
+
+func (r *Reconciler) reconcileJob(ctx context.Context, scheduler *v1alpha1.Scheduler, topic, jobName string) (string, error) {
+	if scheduler.Spec.Project == "" {
+		project, err := metadata.ProjectID()
+		if err != nil {
+			logging.FromContext(ctx).Desugar().Error("Failed to find project id", zap.Error(err))
+			return "", err
+		}
+		scheduler.Spec.Project = project
+	}
```

Comment:
this is wrong everywhere. We shouldn't be "changing" the spec. We are actually not, but this is confusing... 
Should update in every place were we use this metadata server.
Will create a util. Further, this should be mocked.

Reply:
fixed